cmd/zc_traverser_list.go

```diff
@@ -0,0 +1,127 @@
+package cmd
+
+import (
+	"bufio"
+	"fmt"
+	"github.com/Azure/azure-storage-azcopy/common"
+	"net/url"
+	"os"
+)
+
+// a meta traverser that goes through a list of paths (potentially directory entities) and scans them one by one
+// behaves like a single traverser (basically a "traverser of traverser")
+type listTraverser struct {
+	fileLocation            string
+	childTraverserGenerator childTraverserGenerator
+}
+
+type childTraverserGenerator func(childPath string) (resourceTraverser, error)
+
+func (l *listTraverser) traverse(processor objectProcessor, filters []objectFilter) (err error) {
+	f, err := os.Open(l.fileLocation)
+	if err != nil {
+		return fmt.Errorf("unable to open %s to retrieve the required list of entities to transfer", l.fileLocation)
+	}
+	defer f.Close()
+
+	// spawn a scanner to read the list of entities one line at a time
+	scanner := bufio.NewScanner(f)
+	for scanner.Scan() {
+		childPath := scanner.Text()
+
+		// fetch an appropriate traverser, and go through the child path, which could be
+		//   1. a single entity
+		//   2. a directory entity that needs to be scanned
+		childTraverser, err := l.childTraverserGenerator(childPath)
+		if err != nil {
+			glcm.Info(fmt.Sprintf("Skipping %s due to error %s", childPath, err))
+			continue
+		}
+
+		err = childTraverser.traverse(processor, filters)
```

Comment:
Quick note: Traversers return relative paths from the origin. This means that if I supply you a list akin to the following:

FolderA/
FolderB/
FolderC/file.txt

and a source with the following structure:

FolderA/file.txt
FolderA/abc.txt
FolderB/file.txt
FolderB/def.txt
FolderC/file.txt

You will process with the following relative paths:

"/file.txt"
"/abc.txt"
"/file.txt"
"/def.txt"
""

To solve this in my implementation of a glob traverser, I simply added a middleman processor that'd create the ACTUAL relative path using the original base path.

My implementation was roughly as follows. t.GetBasePath() just stripped the wildcards from the base path and used that to construct the relative path from the base.

```go
nt := newLocalTraverser(v, t.recursive, t.incrementEnumerationCounter)
  injectedProcessor := func(object storedObject) error {
   relativePath := strings.TrimPrefix(v, t.getBasePath())
   object.relativePath = relativePath

   return processor(object)
  }

  err = nt.traverse(injectedProcessor, filters)
  if err != nil {
   return err
  }```

Reply:
Pardon me, it seems as though I missed line 72, which seams to clean this up.

Reply:
No worries! I thought of doing the same thing, but I wanted the full path to be present before the filters operate on the stored objects.

I should adopt your implementation though, so that we can better merge our code.